### ISSUE TITLE
检测任务能输出置信度吗

### ISSUE THREAD

**Issue by mrlihellohorld** [IDEA-Research/Rex-Omni#76]
我想做监控场景下的人物性别识别 发现很多误识别，本来是男性误识别为女性。能通过输出置信度规避吗

**Comment by mrlihellohorld** [comment]
https://www.imgur.la/image/vlcsnap-2025-11-17-10h15m14s987.6CBG63
https://www.imgur.la/image/vlcsnap-2025-11-17-10h15m22s164.6CL6of
https://www.imgur.la/image/vlcsnap-2025-11-17-10h15m41s099.6CBo3a
https://www.imgur.la/image/vlcsnap-2025-11-17-10h15m52s858.6CLAIr
https://www.imgur.la/image/vlcsnap-2025-11-17-10h15m59s737.6CLOF5
https://www.imgur.la/image/vlcsnap-2025-11-17-10h17m33s864.6CLnEI
https://www.imgur.la/image/vlcsnap-2025-11-17-10h28m17s886.6CBq2p
https://www.imgur.la/image/vlcsnap-2025-11-17-10h28m21s326.6CBYZO
https://www.imgur.la/image/vlcsnap-2025-11-17-10h32m44s833.6CLa3j
https://www.imgur.la/image/vlcsnap-2025-11-17-10h32m53s506.6CLWWq
https://www.imgur.la/image/vlcsnap-2025-11-17-10h33m21s314.6CLTM0
https://www.imgur.la/image/vlcsnap-2025-11-17-10h15m10s958.6CLrw2
我在恢复中上传不了图片，附上了链接您可以看看。
`#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Basic object detection example using Rex Omni
递归读取test_date_woman/人物属性检测目录下的所有图片，检测woman和person并标注
"""

import os
import time
import torch
from pathlib import Path
from PIL import Image, ImageDraw, ImageFont
from rex_omni import RexOmniVisualize, RexOmniWrapper

os.environ["CUDA_VISIBLE_DEVICES"] = "0"

def get_all_images(dataset_dir):
    """递归获取目录下所有图片文件"""
    image_extensions = {'.jpg', '.jpeg', '.png', '.bmp', '.gif', '.tiff', '.webp'}
    image_files = []
    
    dataset_path = Path(dataset_dir)
    if not dataset_path.exists():
        print(f"错误: 目录 {dataset_dir} 不存在")
        return image_files
    
    # 递归遍历所有文件
    for root, dirs, files in os.walk(dataset_path):
        for file in files:
            file_path = Path(root) / file
            if file_path.suffix.lower() in image_extensions:
                image_files.append(file_path)
    
    return sorted(image_files)


def calculate_iou(box1, box2):
    """计算两个bbox的IoU"""
    x1_1, y1_1, x2_1, y2_1 = box1
    x1_2, y1_2, x2_2, y2_2 = box2
    
    # 计算交集
    x1_i = max(x1_1, x1_2)
    y1_i = max(y1_1, y1_2)
    x2_i = min(x2_1, x2_2)
    y2_i = min(y2_1, y2_2)
    
    if x2_i <= x1_i or y2_i <= y1_i:
        return 0.0
    
    inter_area = (x2_i - x1_i) * (y2_i - y1_i)
    box1_area = (x2_1 - x1_1) * (y2_1 - y1_1)
    box2_area = (x2_2 - x1_2) * (y2_2 - y1_2)
    union_area = box1_area + box2_area - inter_area
    
    if union_area == 0:
        return 0.0
    
    return inter_area / union_area


def filter_overlapping_persons(person_boxes, woman_boxes, iou_threshold=0.5):
    """过滤掉与woman重叠的person boxes"""
    filtered_boxes = []
    for person_box in person_boxes:
        is_overlapping = False
        for woman_box in woman_boxes:
            iou = calculate_iou(person_box, woman_box)
            if iou > iou_threshold:
                is_overlapping = True
                break
        if not is_overlapping:
            filtered_boxes.append(person_box)
    return filtered_boxes


def add_text_to_image(image, text, position=(10, 10), font_size=30):
    """在图片上添加文字"""
    draw = ImageDraw.Draw(image)
    
    # 加载字体
    try:
        font = ImageFont.truetype("/usr/share/fonts/truetype/dejavu/DejaVuSans-Bold.ttf", font_size)
    except:
        try:
            font = ImageFont.truetype("arial.ttf", font_size)
        except:
            font = ImageFont.load_default()
    
    # 获取文字边界框
    bbox = draw.textbbox(position, text, font=font)
    text_width = bbox[2] - bbox[0]
    text_height = bbox[3] - bbox[1]
    
    # 绘制半透明背景
    padding = 10
    background_box = [
        position[0] - padding,
        position[1] - padding,
        position[0] + text_width + padding,
        position[1] + text_height + padding
    ]
    overlay = Image.new('RGBA', image.size, (0, 0, 0, 0))
    overlay_draw = ImageDraw.Draw(overlay)
    overlay_draw.rectangle(background_box, fill=(0, 0, 0, 180))
    image = Image.alpha_composite(image.convert('RGBA'), overlay).convert('RGB')
    draw = ImageDraw.Draw(image)
    
    # 重新加载字体（因为image被转换了）
    try:
        font = ImageFont.truetype("/usr/share/fonts/truetype/dejavu/DejaVuSans-Bold.ttf", font_size)
    except:
        try:
            font = ImageFont.truetype("arial.ttf", font_size)
        except:
            font = ImageFont.load_default()
    
    # 绘制文字（白色，带黑色描边）
    for adj in range(-2, 3):
        for adj2 in range(-2, 3):
            if adj != 0 or adj2 != 0:
                draw.text(
                    (position[0] + adj, position[1] + adj2),
                    text,
                    fill="black",
                    font=font,
                )
    draw.text(position, text, fill="white", font=font)
    
    return image


def main():
    # Model path - replace with your actual model path
    model_path = "models/Rex-Omni-AWQ"
    
    # 测量模型加载时间
    print("正在加载模型...")
    load_start_time = time.time()
    # Create wrapper with custom parameters
    rex_model = RexOmniWrapper(
        model_path=model_path,
        backend="transformers",  # or "vllm" for faster inference  transformers
        max_tokens=512,  # 检测任务：通常 512–1024 已足够，更小可减少生成时间
        temperature=0.0,  # 确定性输出（适合检测、OCR）
        top_p=0.05,
        top_k=1,  # 只从概率最高的 K 个词中选择
        repetition_penalty=1.05,
        torch_dtype=torch.float16,
        #quantization="awq",#vllm模型
    )
    load_end_time = time.time()
    load_time = load_end_time - load_start_time
    print(f"模型加载耗时: {load_time:.2f} 秒")

    # Object detection categories - 同时检测woman和person
    categories = ['woman', 'person']
    
    # 数据集目录
    dataset_dir = "test_date_woman/人物属性检测"
    
    # 输出目录
    output_dir = "/the_shared_storage/lzq/action/video_action/Rex-Omni-master/test_date_woman/人物属性检测/woman_detection_results"
    
    # 创建输出目录
    os.makedirs(output_dir, exist_ok=True)
    
    # 获取所有图片文件
    print(f"\n正在扫描 {dataset_dir} 目录...")
    image_files = get_all_images(dataset_dir)
    total_images = len(image_files)
    
    if total_images == 0:
        print(f"错误: 在 {dataset_dir} 目录下未找到任何图片文件")
        return
    
    print(f"找到 {total_images} 张图片，开始检测...")
    print("-" * 60)
    
    # 统计信息
    error_count = 0
    
    # 处理每张图片
    for idx, image_path in enumerate(image_files, 1):
        try:
            # 打开图片
            image = Image.open(image_path).convert("RGB")
            
            # 进行检测
            results = rex_model.inference(images=image, task="detection", categories=categories)
            
            result = results[0]
            if not result["success"]:
                print(f"[{idx}/{total_images}] {image_path.name}: 检测失败 - {result.get('error', 'Unknown error')}")
                error_count += 1
                continue
            
            predictions = result.get("extracted_predictions", {})
            
            # 提取woman和person的bbox
            woman_boxes = []
            person_boxes = []
            
            if isinstance(predictions, dict):
                # 提取woman boxes
                for category_name in predictions.keys():
                    if category_name.lower().strip() == "woman":
                        detections = predictions[category_name]
                        if isinstance(detections, list):
                            for detection in detections:
                                if isinstance(detection, dict) and detection.get("type") == "box":
                                    coords = detection.get("coords", [])
                                    if coords and len(coords) >= 4:
                                        woman_boxes.append(coords)
                
                # 提取person boxes
                for category_name in predictions.keys():
                    if category_name.lower().strip() == "person":
                        detections = predictions[category_name]
                        if isinstance(detections, list):
                            for detection in detections:
                                if isinstance(detection, dict) and detection.get("type") == "box":
                                    coords = detection.get("coords", [])
                                    if coords and len(coords) >= 4:
                                        person_boxes.append(coords)
            
            # 过滤掉与woman重叠的person boxes（woman也是person，避免重复标注）
            other_person_boxes = filter_overlapping_persons(person_boxes, woman_boxes)
            
            # 统计数量
            woman_count = len(woman_boxes)
            total_person_count = len(person_boxes)  # 总人数（包括woman）
            
            # 构建新的predictions字典，用于可视化
            # woman用红色，其他人用绿色
            vis_predictions = {}
            
            # 添加woman（红色）
            if woman_boxes:
                vis_predictions["woman"] = [
                    {"type": "box", "coords": box} for box in woman_boxes
                ]
            
            # 添加其他人（绿色）
            if other_person_boxes:
                vis_predictions["person"] = [
                    {"type": "box", "coords": box} for box in other_person_boxes
                ]
            
            # 自定义颜色：woman红色，person绿色
            custom_colors = {
                "woman": (255, 0, 0),  # 红色
                "person": (0, 255, 0),  # 绿色
            }
            
            # 可视化
            vis_image = RexOmniVisualize(
                image=image,
                predictions=vis_predictions,
                font_size=20,
                draw_width=3,
                show_labels=True,
                custom_colors=custom_colors,
            )
            
            # 在图片上添加统计信息文字
            text = f""
            vis_image = add_text_to_image(vis_image, text, position=(10, 10), font_size=30)
            
            # 构建相对路径以保持目录结构
            relative_path = image_path.relative_to(Path(dataset_dir))
            output_subdir = relative_path.parent
            
            # 保存结果
            output_path = Path(output_dir) / output_subdir / image_path.name
            output_path.parent.mkdir(parents=True, exist_ok=True)
            vis_image.save(str(output_path))
            
            print(f"[{idx}/{total_images}] {image_path.name}: 女性={woman_count}, 总人数={total_person_count} -> {output_path}")
            
            # 每处理10张图片显示一次进度
            if idx % 10 == 0:
                print(f"  进度: {idx}/{total_images}")
        
        except Exception as e:
            print(f"[{idx}/{total_images}] {image_path.name}: 处理出错 - {str(e)}")
            import traceback
            traceback.print_exc()
            error_count += 1
            continue
    
    # 打印统计结果
    print("\n" + "=" * 60)
    print("检测完成！统计结果：")
    print("=" * 60)
    print(f"总图片数: {total_images}")
    print(f"处理错误: {error_count}")
    print(f"成功处理: {total_images - error_count}")
    print(f"\n结果图片保存在: {output_dir}")
    print("=" * 60)


if __name__ == "__main__":
    main()
`
